File: apps/visualstudio/ResXGuard.Core/Services/LocaleColumns.cs
namespace ResXGuard.Core;

public static class LocaleColumns
{
    public static List<string> MergeVisibleLocales(IReadOnlyList<string> current, IReadOnlyList<string> discovered)
    {
        if (discovered.Count == 0) return new List<string>();
        if (current.Count == 0) return discovered.ToList();
        var known = new HashSet<string>(discovered);
        var kept = current.Where(known.Contains).ToList();
        var added = discovered.Where(l => !current.Contains(l)).ToList();
        var next = kept.Concat(added).ToList();
        return next.Count > 0 ? next : discovered.ToList();
    }
}

File: packages/core-ts/src/services/locale-columns.ts
/** Keep user-visible language columns in sync when .resx files appear or vanish. */
export function mergeVisibleLocales(current: string[], discovered: string[]): string[] {
  if (discovered.length === 0) {
    return [];
  }
  if (current.length === 0) {
    return [...discovered];
  }
  const known = new Set(discovered);
  const kept = current.filter((locale) => known.has(locale));
  const added = discovered.filter((locale) => !current.includes(locale));
  const next = [...kept, ...added];
  return next.length > 0 ? next : [...discovered];
}
